mm2src/coins/qrc20/rpc_clients.rs

```diff
@@ -0,0 +1,336 @@
+    pub gas_used: u64,
+    /// Contract address created, or `None` if not a deployment.
+    #[serde(rename = "contractAddress")]
+    pub contract_address: Option<String>,
+    /// Logs generated within this transaction.
+    pub log: Vec<LogEntry>,
+    /// Whether corresponding contract call (specified in UTXO outputs[output_index]) was failed.
+    /// If None or Some("None") - completed, else failed.
+    pub excepted: Option<String>,
+    #[serde(rename = "exceptedMessage")]
+    pub excepted_message: Option<String>,
+}
+
+#[derive(Debug, Deserialize, PartialEq, Serialize)]
+pub struct LogEntry {
+    /// Contract address.
+    pub address: String,
+    /// Vector of 0x-prefixed hex strings with length of 64.
+    pub topics: Vec<String>,
+    /// In other words the data means a transaction value.
+    pub data: String,
+}
+
+impl LogEntry {
+    pub fn parse_address(&self) -> Result<H160, String> {
+        if self.address.starts_with("0x") {
+            qtum::contract_addr_from_str(&self.address)
+        } else {
+            let address = format!("0x{}", self.address);
+            qtum::contract_addr_from_str(&address)
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum TopicFilter {
+    Match(String),
+    Skip,
+}
+
+impl From<&str> for TopicFilter {
+    fn from(topic: &str) -> Self { TopicFilter::Match(topic.to_string()) }
+}
+
+/// Qrc20 specific RPC ops
+pub trait Qrc20RpcOps {
+    /// This can be used to get the basic information(name, decimals, total_supply, symbol) of a QRC20 token.
+    /// https://github.com/qtumproject/qtum-electrumx-server/blob/master/docs/qrc20-integration.md#blockchaintokenget_infotoken_address
+    fn blockchain_token_get_info(&self, token_addr: &H160Json) -> RpcRes<TokenInfo>;
+
+    fn blockchain_contract_call(&self, contract_addr: &H160Json, data: BytesJson) -> RpcRes<ContractCallResult>;
+
+    /// This can be used to retrieve QRC20 token transfer history, params are the same as blockchain.contract.event.subscribe,
+    /// and it returns a list of map{tx_hash, height, log_index}, where log_index is the position for this event log in its transaction.
+    /// https://github.com/qtumproject/qtum-electrumx-server/blob/master/docs/qrc20-integration.md#blockchaincontracteventget_historyhash160-contract_addr-topic
+    fn blockchain_contract_event_get_history(
+        &self,
+        address: &H160Json,
+        contract_addr: &H160Json,
+        topic: &str,
+    ) -> RpcRes<Vec<TxHistoryItem>>;
+
+    /// This can be used to get eventlogs in the transaction, the returned data is the same as Qtum Core RPC gettransactionreceipt.
+    /// from the eventlogs, we can get QRC20 Token transafer informations(from, to, amount).
+    /// https://github.com/qtumproject/qtum-electrumx-server/blob/master/docs/qrc20-integration.md#blochchaintransactionget_receipttxid
+    fn blochchain_transaction_get_receipt(&self, hash: &H256Json) -> RpcRes<Vec<TxReceipt>>;
+}
+
+pub trait Qrc20NativeOps {
```

Comment:
Same here as for `Qrc20NativeWalletOps`.

Reply:
Small course change - looks like methods from this trait can be "merged" to the Qrc20RpcOps. The `get_addresses_by_label` is not QTUM only (e.g. https://developer.bitcoin.org/reference/rpc/getaddressesbylabel.html), so it can be moved to the NativeClientImpl.

Reply:
Done